File: packages/flutter/test/redaction_fixtures_test.dart
import 'dart:convert';
import 'dart:io';

import 'package:crumbtrail_flutter/crumbtrail_flutter.dart';
import 'package:flutter_test/flutter_test.dart';

/// Conformance against `test-fixtures/redaction/urls.json`.
///
/// The Kotlin SDK runs the equivalent of this file against the same corpus, so
/// a rule that drifts in one language fails in both. The class doc on
/// `CrumbtrailRedaction` claims the SDKs are held to the same fixtures; this is
/// the file that makes that true for URLs.
///
/// The corpus is read from the repo root rather than copied in: a per-SDK copy
/// would hide exactly the cross-language drift it exists to catch.
void main() {
  final fixture = () {
    var dir = Directory.current;
    while (!Directory('${dir.path}/test-fixtures/redaction').existsSync()) {
      final parent = dir.parent;
      if (parent.path == dir.path) {
        throw StateError('repo root not found from ${Directory.current.path}');
      }
      dir = parent;
    }
    return File('${dir.path}/test-fixtures/redaction/urls.json');
  }();

  List<Map<String, String>> cases() {
    final root = jsonDecode(fixture.readAsStringSync()) as Map<String, Object?>;
    return (root['cases'] as List<Object?>)
        .cast<Map<String, Object?>>()
        .map((entry) => entry.map((key, value) => MapEntry(key, '$value')))
        .toList();
  }

  test('the corpus is reachable and not empty', () {
    // If the path arithmetic above is wrong every other assertion here would
    // pass vacuously against an empty list. Fail loudly instead.
    expect(fixture.existsSync(), isTrue, reason: 'expected ${fixture.path}');
    expect(cases().length, greaterThanOrEqualTo(10));
  });

  test('every fixture URL redacts to the shared expectation', () {
    final failures = <String>[];
    for (final entry in cases()) {
      final actual = CrumbtrailRedaction.redactUrl(entry['input']!);
      if (actual != entry['expected']) {
        failures.add('${entry['name']}\n'
            '  input:    ${entry['input']}\n'
            '  expected: ${entry['expected']}\n'
            '  actual:   $actual');
      }
    }
    expect(failures, isEmpty,
        reason: 'test-fixtures/redaction/urls.json disagrees with this SDK:\n'
            '${failures.join('\n')}');
  });

  test('the marker is emitted literally so ingest can match on it', () {
    // Re-encoding the marker is the failure mode that hides a redaction from
    // every downstream consumer while still looking redacted to a reader.
    expect(
      CrumbtrailRedaction.redactUrl('https://api.example.com/api?token=abc'),
      'https://api.example.com/api?token=[REDACTED]',
    );
  });

  test('verified leaks from the reviewed adapters stay redacted', () {
    const urls = [
      'https://api.example.com/v1/users/omar@shabana.dev/profile',
      'https://api.example.com/reset-password/eyJhbGciOiJIUzI1NiJ9.abc.def',
      'https://api.example.com/api?q=Bearer%20sk-live-1234567890',
      'https://api.example.com/api?next=https%3A%2F%2Fx.com%3Ftoken%3Dsecret',
    ];
    const secrets = [
      'omar@shabana.dev',
      'eyJhbGciOiJIUzI1NiJ9',
      'sk-live-1234567890',
      'token%3Dsecret',
    ];
    for (final url in urls) {
      final redacted = CrumbtrailRedaction.redactUrl(url);
      expect(redacted, contains(CrumbtrailRedaction.placeholder));
      for (final secret in secrets) {
        expect(redacted, isNot(contains(secret)));
      }
    }
  });
}

File: packages/kotlin/src/test/kotlin/ai/crumbtrail/sdk/RedactionFixturesTest.kt
package ai.crumbtrail.sdk

import java.io.File
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertTrue

/**
 * Conformance against `test-fixtures/redaction/urls.json`.
 *
 * The Dart SDK runs the equivalent of this file against the same corpus, so a
 * rule that drifts in one language fails in both. The class doc on
 * [CrumbtrailRedaction] claims the SDKs are held to the same fixtures; this is
 * the file that makes that true for URLs.
 *
 * The corpus is read from the repo root rather than copied in: a per-SDK copy
 * would hide exactly the cross-language drift it exists to catch.
 */
class RedactionFixturesTest {
    private val fixture: File = run {
        var dir = File(System.getProperty("user.dir"))
        while (!File(dir, "test-fixtures/redaction").isDirectory) {
            dir = dir.parentFile ?: error("repo root not found from ${System.getProperty("user.dir")}")
        }
        File(dir, "test-fixtures/redaction/urls.json")
    }

    private fun cases(): List<Triple<String, String, String>> {
        val root = MiniJson(fixture.readText()).parseValue() as JsonValue.Obj
        val list = root.values["cases"] as JsonValue.Arr
        return list.values.map { entry ->
            val case = entry as JsonValue.Obj
            fun field(name: String) = (case.values[name] as JsonValue.Str).value
            Triple(field("name"), field("input"), field("expected"))
        }
    }

    @Test
    fun `the corpus is reachable and not empty`() {
        // If the path arithmetic above is wrong every other assertion here would
        // pass vacuously against an empty list. Fail loudly instead.
        assertTrue(fixture.isFile, "expected ${fixture.absolutePath} to exist")
        assertTrue(cases().size >= 10, "expected a corpus, found ${cases().size} cases")
    }

    @Test
    fun `every fixture URL redacts to the shared expectation`() {
        val failures = cases().mapNotNull { (name, input, expected) ->
            val actual = CrumbtrailRedaction.redactUrl(input)
            if (actual == expected) null else "$name\n  input:    $input\n  expected: $expected\n  actual:   $actual"
        }
        assertTrue(
            failures.isEmpty(),
            "test-fixtures/redaction/urls.json disagrees with this SDK:\n" + failures.joinToString("\n"),
        )
    }

    @Test
    fun `the marker is emitted literally so ingest can match on it`() {
        // Re-encoding the marker is the failure mode that hides a redaction from
        // every downstream consumer while still looking redacted to a reader.
        val redacted = CrumbtrailRedaction.redactUrl("https://api.example.com/api?token=abc")
        assertEquals("https://api.example.com/api?token=[REDACTED]", redacted)
    }

    @Test
    fun `verified leaks from the reviewed adapters stay redacted`() {
        for (url in listOf(
            "https://api.example.com/v1/users/omar@shabana.dev/profile",
            "https://api.example.com/reset-password/eyJhbGciOiJIUzI1NiJ9.abc.def",
            "https://api.example.com/api?q=Bearer%20sk-live-1234567890",
            "https://api.example.com/api?next=https%3A%2F%2Fx.com%3Ftoken%3Dsecret",
        )) {
            val redacted = CrumbtrailRedaction.redactUrl(url)
            assertTrue(redacted.contains(CrumbtrailRedaction.PLACEHOLDER), "not redacted: $redacted")
            for (secret in listOf("omar@shabana.dev", "eyJhbGciOiJIUzI1NiJ9", "sk-live-1234567890", "token%3Dsecret")) {
                assertTrue(!redacted.contains(secret), "leaked $secret in $redacted")
            }
        }
    }
}
